theon/src/allocator.rs

```diff
@@ -94,15 +481,28 @@ mod global {
         }
     }
 
-    unsafe impl<T: Heap> GlobalAlloc for BumpAlloc<T> {
+    /// GlobalQuickAlloc is a wrapper around a QuickFit over a
+    /// GlobalHeap that uses interior mutability to implement
+    /// the GlobalAlloc trait.
+    struct GlobalQuickAlloc(UnsafeCell<QuickFit<GlobalHeap>>);
```

Comment:
I think the introduction of `AtomicPtr` to guard against inadvertent concurrent access should suffice here? Since I `swap` the pointer with NULL, callers will either see a valid pointer to the QuickFit allocator, or they'll see nothing (and like it!). Since `theon` is single-threaded by construction, that should be good to go.

Reply:
Yeah, this should work for the single threaded context, but how about wrapping QuickFit in a spin lock rather than an AtomicPtr? In a single threaded context it's not going to have any contention. It should be one atomic op to both acquire and release, which matches the swaps on your AtomicPtr but with a better understood primitive.

Reply:
Mind if I push that off to a future revision? Hypatia as a whole will need locks eventually, but they are not yet and I'd rather not drag in `spin` (though admittedly more out of a general desire to minimize dependencies, rather than any particular concern about soundness or what not).

And honestly I'd really rather that we unabashedly crash if we ever encounter a situation where there are concurrent allocator calls in Theon.

Reply:
Sadly, it just doesn't do too much right now, so perhaps not that surprising. :-/